ksqldb-engine/src/main/java/io/confluent/ksql/serde/InternalFormats.java

```diff
@@ -50,16 +50,16 @@ private InternalFormats() {
    * @see <a href=https://github.com/confluentinc/ksql/issues/6296>Issue 6296</a>
    * @see SerdeFeaturesFactory#buildInternal
    */
-  public static Formats of(final FormatInfo keyFormat, final FormatInfo valueFormat) {
+  public static Formats of(final KeyFormat keyFormat, final FormatInfo valueFormat) {
     // Do not use NONE format for internal topics:
-    if (keyFormat.getFormat().equals(NoneFormat.NAME)) {
+    if (keyFormat.getFormatInfo().getFormat().equals(NoneFormat.NAME)) {
       throw new IllegalArgumentException(NoneFormat.NAME + " can not be used for internal topics");
     }
 
     return Formats.of(
-        keyFormat,
+        keyFormat.getFormatInfo(),
         valueFormat,
-        SerdeFeaturesFactory.buildInternal(FormatFactory.fromName(keyFormat.getFormat())),
+        keyFormat.getFeatures(),
```

Comment:
It's odd to me that we're wiping value features on internal topics while passing along key features. I need to think more about the implications here. We might need to write key serde features into all the execution steps if we want this to be evolvable in the future.

Reply:
Took a deeper look -- looks like we already write serde features into the execution steps that need them, so that shouldn't be a concern. 

The only potential evolvability concern is that until now, our execution plans have always specified key wrapping in ambiguous situations (single key column with format that supports both wrapping and unwrapping). That changes with this PR, which means we'd be locking ourselves into the current default choices for each format -- namely, always wrap by default. Are we happy with this default, or should we consider defaulting key wrapping to unwrap even though value wrapping default wrapping? If most users unwrap single keys, we may prefer the former despite the inconsistency.

Reply:
> What changes with this PR, which means we'd be locking ourselves into the current default choices for each format -- namely, always wrap by default.

Why does it mean that? If we write the serde features into the execution steps then we can always change the logical plan to produce different execution steps while keeping all of the old ones compatible. 

> If most users unwrap single keys, we may prefer the former despite the inconsistency.

Users shouldn't see what happens underneath the hood, we still "inherit" the desired serde features to the sink independent of what happens in the internal topics. You can see the example I've given in the "testing done" where the table is wrapped and the stream is unwrapped. The "final" chooses the left topic's information, which is to `UNWRAP_SINGLES`

Reply:
> It's odd to me that we're wiping value features on internal topics while passing along key features. 

I agree, it felt a little sticky to me as well. I think it's possible to pass along the value format as well, but it's not as interesting (it would be mostly for hygiene) because we can always serialize it into the desired format at the edges (our semantics don't depend on the value format).

If we do this, we can do it as a follow-up.